sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -1308,23 +1308,27 @@ std::unique_ptr<CurlNetworkConnection> CurlConnectionPool::ExtractOrCreateCurlCo
         + std::string("curl_easy_init returned Null"));
   }
   CURLcode result;
-
+  auto curlConnection = std::make_unique<CurlConnection>(newHandle, connectionKey);
```

Comment:
Maintenance nit: add the explanation in source comment, explaining why the unique ptr needs to be created here, to manage lifetime properly, before a throw. This way, figure code changes do not accidentally re-introduce the bug.

Reply:
Also, any easy way to add a test to ensure this doesn't regress?

Reply:
Actually, it might be easier to read/follow/understand if we add a `curl_easy_cleanup(newHandle)` for each throw.

Rather than creating an instance of a class that uses the destructor to call the cleanup.  
People might miss the comment that mentions this.  So, a little duplication might be not that bad, with the benefit of quickly seeing what's happening there.  And the Connection instance is not even created on throwing.

Somehow we are relying on a destructor of CurlConnection to handle the leak, but in the future, this might be easy to forget, as @ahsonkhan  mentioned

Reply:
Agree with Victor, i went the add curl easy cleanup, even though it is more noisy from a code change persp[ective, makes it more clear as to what is going on

Reply:
Piling onto my comment: In C++ code, you have two choices:
1) Disallow the use of all exceptions and allow raw pointers and handles.
1) Allow exceptions and use RAII types for all pointers and handles.

Those are the *only* two options available for C++ code. Any attempts to do anything other than that are pretty much guaranteed to leak memory or have other issues.

And because I am addicted to the C++ Core Guidelines, I'll throw out "[Don't Leak Any Resources](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#p8-dont-leak-any-resources)".

Reply:
i fully agree with you @LarryOsterman ,